### DISCUSSION TITLE
Process for setting up immich on shared postgres without super user?

### DISCUSSION THREAD

**Discussion by cbc02009** [immich-app/immich#8327] [Q&A] [answered]
Hello,

Is there any documentation around setting up a new immich instance that uses a shared monolithic postgres db and doesn't have superuser privileges? I've been trying PG commands from changelog notes and error messages from immich microservices, but I still keep getting errors and I think it may be because I'm not running the right commands or running them as the correct user.

If it's helpful, I'm setting up immich on a flux-controlled k8s cluster. The immich setup is available [here](https://github.com/cbc02009/k8s-home-ops/tree/main/kubernetes/main/apps/selfhosted/immich/app)

I have run the following commands as a superuser as described in the error messages:
```
CREATE EXTENSION IF NOT EXISTS vectors;
CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE;
```
and now I am getting the following error and don't know how to proceed:
```
[Nest] 7  - 03/28/2024, 12:21:52 PM     LOG [CommunicationRepository] Initialized websocket server
Migration "UsePgVectors1700713871511" failed, error: type "vector" does not exist
/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219
            throw new QueryFailedError_1.QueryFailedError(query, parameters, err);
                  ^

QueryFailedError: type "vector" does not exist
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async UsePgVectors1700713871511.up (/usr/src/app/dist/infra/migrations/1700713871511-UsePgVectors.js:19:9)
    at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
    at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
    at async DatabaseRepository.runMigrations (/usr/src/app/dist/infra/repositories/database.repository.js:178:9)
    at async /usr/src/app/dist/domain/database/database.service.js:52:13
    at async /usr/src/app/dist/infra/repositories/database.repository.js:186:23 {
  query: '\n' +
    '        ALTER TABLE asset_faces \n' +
    '        ALTER COLUMN embedding SET NOT NULL,\n' +
    '        ALTER COLUMN embedding TYPE vector(512)',
  parameters: undefined,
  driverError: error: type "vector" does not exist
      at /usr/src/app/node_modules/pg/lib/client.js:526:17
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:184:25)
      at async UsePgVectors1700713871511.up (/usr/src/app/dist/infra/migrations/1700713871511-UsePgVectors.js:19:9)
      at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
      at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:265:35)
      at async DatabaseRepository.runMigrations (/usr/src/app/dist/infra/repositories/database.repository.js:178:9)
      at async /usr/src/app/dist/domain/database/database.service.js:52:13
      at async /usr/src/app/dist/infra/repositories/database.repository.js:186:23 {
    length: 89,
    severity: 'ERROR',
    code: '42704',
    detail: undefined,
    hint: undefined,
    position: undefined,
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: undefined,
    table: undefined,
    column: undefined,
    dataType: undefined,
    constraint: undefined,
    file: 'parse_type.c',
    line: '270',
    routine: 'typenameType'
  },
  length: 89,
  severity: 'ERROR',
  code: '42704',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'parse_type.c',
  line: '270',
  routine: 'typenameType'
}
```

Any help would be greatly appreciated, thanks!

**Comment by mmomjian** [comment]
Please connect to the immich DB and run `\dx` and `\dz`, post  the results.

**Reply by cbc02009** [reply]
sorry for the (hopefully not too dumb) question. I've only ever messed with the DB using PGAdmin. how do you want me to connect to the immich db and run those commands? from inside one of the immich containers?

**Reply by mmomjian** [reply]
You need to run it with the PG console, same way you ran create extension

Running on a monolithic DB does require some baseline level of facility with the `psql` console, though.

**Reply by cbc02009** [reply]
output of `\dx` is 
```
                                                    List of installed extensions
     Name      | Version |   Schema   |                                         Description

---------------+---------+------------+----------------------------------------------------------------------------------------------
 cube          | 1.5     | public     | data type for multidimensional cubes
 earthdistance | 1.1     | public     | calculate great-circle distances on the surface of the Earth
 plpgsql       | 1.0     | pg_catalog | PL/pgSQL procedural language
 uuid-ossp     | 1.1     | public     | generate universally unique identifiers (UUIDs)
 vectors       | 0.2.1   | vectors    | vectors: Vector database plugin for Postgres, written in Rust, specifically designed for LLM
```
`\dz`  is an invalid command

**Reply by mmomjian** [reply]
Sorry, it's `\z`. Can you also run `show search_path;`

Can you please post the entire output including the line you wrote the command on? I want to check which DB you are connected to.

**Reply by cbc02009** [reply]
Sure, the missing lines are:
```
postgres=# \c immich
You are now connected to database "immich" as user "postgres".
immich=# \dx
```

output of `\z` is:
```
immich=# \z
                                    Access privileges
 Schema |       Name        |   Type   | Access privileges | Column privileges | Policies
--------+-------------------+----------+-------------------+-------------------+----------
 public | migrations        | table    |                   |                   |
 public | migrations_id_seq | sequence |                   |                   |
 public | typeorm_metadata  | table    |                   |                   |
(3 rows)
```

output of `show search_path;`:
```
immich=# show search_path;
   search_path
-----------------
 "$user", public
(1 row)
```

**Reply by mmomjian** [reply]
As @bo0tzz alluded to, try running `ALTER DATABASE immich SET search_path TO "$user", public, vectors;`

We might also need to run `GRANT SELECT ON SCHEMA` but that's not the current issue. I *think* they may have fixed vecto.rs 0.21 to do this automatically, so I'd be curious to see if it works for you with just the first one.

**Reply by cbc02009** [reply]
Same exact error after  `ALTER DATABASE immich SET search_path TO "$user", public, vectors;`

`GRANT SELECT ON SCHEMA` also does not change the error.

**Reply by mmomjian** [reply]
The full command would be `GRANT SELECT ON SCHEMA vectors TO public;`. However, if the error is exactly the same, I wouldn't expect that to help. I'm surprised that setting the search path didn't change the error at all.

Try also running `SET search_path TO "$user", public, vectors;`. See if that helps. If not, restart postgres cluster and try once more please.

**Reply by mertalev** [reply]
Try `ALTER SCHEMA vectors OWNER TO immich;`.

**Reply by mmomjian** [reply]
@mertalev not sure if this interests you but my DB is running fine with the `vectors` schema owned by postgres. I did run `GRANT ALL ON SCHEMA vectors TO immich;` but I'm not sure if this is needed anymore either.

**Reply by cbc02009** [reply]
After running `ALTER SCHEMA vectors OWNER TO "immich-user";` there's a new error:
```
[Nest] 7  - 03/28/2024, 2:16:26 PM     LOG [CommunicationRepository] Initialized websocket server
[Nest] 7  - 03/28/2024, 2:16:27 PM    WARN [DatabaseService] Could not run vector reindexing checks. If the extension was updated, please restart the Postgres instance.
/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219
            throw new QueryFailedError_1.QueryFailedError(query, parameters, err);
                  ^

QueryFailedError: permission denied for view pg_vector_index_stat
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:219:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async DataSource.query (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:350:20)
    at async DatabaseRepository.shouldReindex (/usr/src/app/dist/infra/repositories/database.repository.js:128:25)
    at async /usr/src/app/dist/domain/database/database.service.js:41:21
    at async /usr/src/app/dist/infra/repositories/database.repository.js:186:23 {
  query: '\n' +
    '          SELECT idx_status\n' +
    '          FROM pg_vector_index_stat\n' +
    '          WHERE indexname = $1',
  parameters: [ 'clip_index' ],
  driverError: error: permission denied for view pg_vector_index_stat
      at /usr/src/app/node_modules/pg/lib/client.js:526:17
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:184:25)
      at async DataSource.query (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:350:20)
      at async DatabaseRepository.shouldReindex (/usr/src/app/dist/infra/repositories/database.repository.js:128:25)
      at async /usr/src/app/dist/domain/database/database.service.js:41:21
      at async /usr/src/app/dist/infra/repositories/database.repository.js:186:23 {
    length: 107,
    severity: 'ERROR',
    code: '42501',
    detail: undefined,
    hint: undefined,
    position: undefined,
    internalPosition: undefined,
    internalQuery: undefined,
    where: undefined,
    schema: undefined,
    table: undefined,
    column: undefined,
    dataType: undefined,
    constraint: undefined,
    file: 'aclchk.c',
    line: '2807',
    routine: 'aclcheck_error'
  },
  length: 107,
  severity: 'ERROR',
  code: '42501',
  detail: undefined,
  hint: undefined,
  position: undefined,
  internalPosition: undefined,
  internalQuery: undefined,
  where: undefined,
  schema: undefined,
  table: undefined,
  column: undefined,
  dataType: undefined,
  constraint: undefined,
  file: 'aclchk.c',
  line: '2807',
  routine: 'aclcheck_error'
}

Node.js v20.11.1
```
Not sure if that made things better or worse.

**Reply by mmomjian** [reply]
Try this one next: `GRANT SELECT ON SCHEMA vectors TO public;`

**Reply by cbc02009** [reply]
It doesn't seem to like that command.
```
immich=# GRANT SELECT ON SCHEMA vectors TO public;
ERROR:  invalid privilege type SELECT for schema
```

**Reply by mmomjian** [reply]
Sorry, my memory is failing me. Refer to my earlier comment from the 1.95 release: https://github.com/immich-app/immich/discussions/7252#discussioncomment-8534336

try 
`GRANT ALL ON SCHEMA vectors TO <immich username>;`

please let us know if that works. If not, try  `GRANT SELECT ON TABLE pg_vector_index_stat to <immich username>;`

**Reply by cbc02009** [reply, marked as answer]
Thanks! 
```
GRANT ALL ON SCHEMA vectors TO <immich username>;
GRANT SELECT ON TABLE pg_vector_index_stat to <immich username>;
```
finally did the trick!

So the final full list of commands is:
```
CREATE EXTENSION IF NOT EXISTS vectors;
CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE;
ALTER DATABASE immich SET search_path TO "$user", public, vectors;
GRANT ALL ON SCHEMA vectors TO <immich username>;
GRANT SELECT ON TABLE pg_vector_index_stat to <immich username>;
```
It's possible one or more of them are unnecessary, but this worked for me.

**Reply by mmomjian** [reply]
Great! I think the unnecessary one is `ALTER SCHEMA vectors OWNER TO <immich username>;` - this is covered by GRANT ALL.

**Reply by cbc02009** [reply]
Thanks, I will edit that one out

**Reply by mertalev** [reply]
`GRANT USAGE` is probably what was missing before owning the schema.

**Comment by bo0tzz** [comment]
Some of the other things that may be necessary for running Immich without postgres superuser:
Setting the search_path
Appropriate permissions (or ownership?) on the vectors schema

See also #7431

@mertalev I'm hoping that we can figure out exactly what is necessary here so we can properly document it. Do you have any other things to add?

**Reply by mmomjian** [reply]
I think my write up from 1.95.0 release is still accurate: https://github.com/immich-app/immich/discussions/7252#discussioncomment-8534336

**Reply by mmomjian** [reply]
So it looks like the final list of commands (at the moment) is:

```
\c <immich db>
ALTER DATABASE <immich db> OWNER TO <immich username>;
CREATE EXTENSION IF NOT EXISTS vectors;
CREATE EXTENSION IF NOT EXISTS earthdistance CASCADE;
ALTER DATABASE <immich db> SET search_path TO "$user", public, vectors;
GRANT USAGE ON SCHEMA vectors TO <immich username>;
GRANT SELECT ON TABLE pg_vector_index_stat to <immich username>;
```

We *may* need to start the Immich server before running the last command? I'm unsure if `pg_vector_index_stat` exists before the vecto.rs tables are created.

**Reply by mertalev** [reply]
I'm not sure either. Would `GRANT SELECT ON ALL TABLES IN SCHEMA vectors` take care of it if it doesn't exist yet?

**Reply by mmomjian** [reply]
I'm thinking `ALTER DEFAULT PRIVILEGES IN SCHEMA vectors GRANT SELECT ON TABLES TO PUBLIC;`? [link](https://www.postgresql.org/docs/current/sql-alterdefaultprivileges.html)

**Reply by mertalev** [reply]
That looks good, but it would only apply to new tables.

Actually, I think this table would exist after creating the extension since we query it before any migrations are run.

**Reply by mmomjian** [reply]
Exactly, it would only apply moving forward. This is the source of my confusion as I am not sure exactly when that table exists - if it indeed exists as soon as the vecto.rs extension is added, the `GRANT SELECT ON TABLE pg_vector_index_stat to <immich username>;` should work. Just hard to test. Lol
